{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework 1\n",
    "### COSC 594 Bioinformatics\n",
    "### Started 01/15/18, Due 01/23/18 \n",
    "### Elliot Greenlee\n",
    "\n",
    "1\\. On my honor, I agree to the terms of the Academic Code of Honor.\n",
    "\n",
    "2\\. See included file, lambda.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input a genome from a file in fasta format\n",
    "def input_genome(filename, chosen_gen):    \n",
    "    with open(filename, 'r') as f:\n",
    "        header = f.readline()\n",
    "        current_gen = 1\n",
    "        genome = \"\"\n",
    "        for line in f:\n",
    "            if line[0] is '>':\n",
    "                if current_gen is chosen_gen:\n",
    "                    genome = genome.replace('\\n', '')\n",
    "                    genome = genome.replace('N', 'A')\n",
    "                    return header, genome\n",
    "                else:\n",
    "                    header = line\n",
    "                    current_gen += 1\n",
    "            else:\n",
    "                genome = genome + line\n",
    "          \n",
    "        if current_gen is chosen_gen:\n",
    "            genome = genome.replace('\\n', '')\n",
    "            genome = genome.replace('N', 'A')\n",
    "            return header, genome\n",
    "        else:\n",
    "            print(\"This file only has {} genomes\".format(current_gen))\n",
    "            exit(1)\n",
    "\n",
    "# Output a genome in fasta format\n",
    "def output_genome(filename, header, genome):\n",
    "    with open(filename, 'w') as f:\n",
    "        f.write(header)\n",
    "        f.write(genome)\n",
    "        f.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. Reverse complement the lambda genome. The results can be found in lambda.rev.fasta\n",
    "# Reverse a genome\n",
    "def reverse(genome):\n",
    "    rev_genome = genome[::-1]\n",
    "    return rev_genome\n",
    "\n",
    "# Complement a genome\n",
    "def complement(genome):\n",
    "    tmp_genome = genome.replace('A', '1')\n",
    "    tmp_genome = tmp_genome.replace('T', '2')\n",
    "    tmp_genome = tmp_genome.replace('C', '3')\n",
    "    tmp_genome = tmp_genome.replace('G', '4')\n",
    "    \n",
    "    comp_genome = tmp_genome.replace('1', 'T')\n",
    "    comp_genome = comp_genome.replace('2', 'A')\n",
    "    comp_genome = comp_genome.replace('3', 'G')\n",
    "    comp_genome = comp_genome.replace('4', 'C')\n",
    "    \n",
    "    return comp_genome\n",
    "\n",
    "# Input the lambda genome\n",
    "header, genome = input_genome(\"lambda.fasta\", 1)\n",
    "\n",
    "# Reverse complement the lambda genome\n",
    "rev_comp_genome = reverse(complement(genome))\n",
    "\n",
    "# Output the reverse complement of the lambda genome\n",
    "output_genome(\"lambda.rev.fasta\", \"> reversed\", rev_comp_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nucleotide Frequencies\n",
      "{'A': 0.2542987918023999, 'T': 0.24712382994515691, 'C': 0.23425838109768668, 'G': 0.2643189971547565}\n",
      "\n",
      "Dinucleotide Frequencies\n",
      "AA 0.07612214181150904\n",
      "AT 0.06880270509886394\n",
      "AC 0.053050452567988286\n",
      "AG 0.056328735489989894\n",
      "TA 0.04474134553926723\n",
      "TT 0.06896765015154327\n",
      "TC 0.05519473825281953\n",
      "TG 0.07822519123317045\n",
      "CA 0.06630791117708913\n",
      "CT 0.05228758169934641\n",
      "CC 0.05148347456753469\n",
      "CG 0.06418424362384281\n",
      "GA 0.06713263644048577\n",
      "GT 0.057070988227046864\n",
      "GC 0.07453454567947053\n",
      "GG 0.06556565844003216\n"
     ]
    }
   ],
   "source": [
    "# 4. Report the nucleotide frequencies and the dinucleotide frequencies of lambda\n",
    "import numpy as np\n",
    "import regex as re\n",
    "\n",
    "nucleotides = ['A', 'T', 'C', 'G']\n",
    "\n",
    "# Compute nucleotide frequencies for a genome\n",
    "def freq_nucleotides(genome):\n",
    "    length = len(genome) * 1.0\n",
    "    frequencies = {}\n",
    "    for nucleotide in nucleotides:\n",
    "        frequencies[nucleotide] = genome.count(nucleotide) / length\n",
    "    \n",
    "    return frequencies\n",
    "\n",
    "# Compute dinucleotide frequencies for a genome\n",
    "def freq_dinucleotides(genome):\n",
    "    num_dinucleotides = len(genome) - 1.0\n",
    "    \n",
    "    frequencies = {}\n",
    "    # Iterate over dinucleotides\n",
    "    for nucleotide1 in nucleotides:\n",
    "        for nucleotide2 in nucleotides:\n",
    "            dinucleotide = nucleotide1 + nucleotide2\n",
    "            frequencies[dinucleotide] = len(re.findall(dinucleotide, genome, overlapped=True)) / num_dinucleotides\n",
    "    \n",
    "    return frequencies\n",
    "\n",
    "# Input the lambda genome\n",
    "header, genome = input_genome(\"lambda.fasta\", 1)\n",
    "\n",
    "# Compute the nucleotide frequencies of the lambda genome\n",
    "print(\"Nucleotide Frequencies\")\n",
    "print(freq_nucleotides(genome))\n",
    "print()\n",
    "\n",
    "# Compute the dinucleotide frequencies\n",
    "print(\"Dinucleotide Frequencies\")\n",
    "frequencies = freq_dinucleotides(genome)\n",
    "for dinucleotide in frequencies:\n",
    "    print(dinucleotide, frequencies[dinucleotide])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5\\. See included files, human_mito.fasta and neander_sample.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Multinomial Probability\n",
      "-452.6623122326945\n",
      "\n",
      "Order 3 Markov Probability\n",
      "-451.7966057821607\n"
     ]
    }
   ],
   "source": [
    "# 6. Compute the log of the probability of the Neanderthal sequence \n",
    "# for a multinomial model and an order 3 markov model, both trained on \n",
    "# the human mitochondrial genome\n",
    "import math\n",
    "import numpy as np\n",
    "\n",
    "nucleotides = ['A', 'T', 'C', 'G']\n",
    "\n",
    "def log_multinomial_prob(training_genome, testing_genome):\n",
    "    # Calculate the nucleotide frequencies from training\n",
    "    nucleotide_frequencies = freq_nucleotides(training_genome)\n",
    "    \n",
    "    # Calculate testing sample log probability\n",
    "    total_probability = 0\n",
    "    for nucleotide in nucleotides:\n",
    "        total_probability += testing_genome.count(nucleotide) * math.log(nucleotide_frequencies[nucleotide])\n",
    "                \n",
    "    return total_probability\n",
    "\n",
    "def log_markov_prob(training_genome, testing_genome):\n",
    "    order = 3\n",
    "    \n",
    "    # Calculate the nucleotide frequencies from training\n",
    "    nucleotide_frequencies = freq_nucleotides(training_genome)\n",
    "    \n",
    "    # Create dictionary of the transitions for all possible 4 letter combinations\n",
    "    transitions = {}\n",
    "    for nucleotide1 in nucleotides:\n",
    "        for nucleotide2 in nucleotides:\n",
    "            for nucleotide3 in nucleotides:\n",
    "                t_sum = 0.0\n",
    "                tmp = {}\n",
    "                for nucleotide4 in nucleotides:\n",
    "                    seq = nucleotide1 + nucleotide2 + nucleotide3 + nucleotide4\n",
    "                    # Find all occurences of this 4 nucleotide sequence\n",
    "                    count = len(re.findall(seq, training_genome, overlapped=True))\n",
    "                    tmp[seq] = count\n",
    "            \n",
    "                    t_sum += count\n",
    "                \n",
    "                # Find transition value by dividing by sum of transitions for this final letter\n",
    "                for key in tmp:\n",
    "                    transitions[key] = tmp[key] / t_sum\n",
    "    \n",
    "    # Optionally print all transition values\n",
    "    #for key in transitions:\n",
    "        #print(key, transitions[key])\n",
    "        \n",
    "    counts = {}\n",
    "    for nucleotide1 in nucleotides:\n",
    "        for nucleotide2 in nucleotides:\n",
    "            for nucleotide3 in nucleotides:\n",
    "                for nucleotide4 in nucleotides:\n",
    "                    seq = nucleotide1 + nucleotide2 + nucleotide3 + nucleotide4\n",
    "                    counts[seq] = len(re.findall(seq, testing_genome, overlapped=True))\n",
    "                    \n",
    "                    \n",
    "    total_probability = 0\n",
    "    \n",
    "    for nucleotide in testing_genome[0:order]:\n",
    "        total_probability += math.log(nucleotide_frequencies[nucleotide])\n",
    "        \n",
    "    for seq in transitions:\n",
    "        total_probability += counts[seq] * math.log(transitions[seq])\n",
    "        \n",
    "    return total_probability\n",
    "\n",
    "# Input the human mitochondrial genome for training and the neanderthal genome for testing\n",
    "training_header, training_genome = input_genome(\"human_mito.fasta\", 1)\n",
    "testing_header, testing_genome = input_genome(\"neander_sample.fasta\", 1)\n",
    "\n",
    "# Calculate the log of the multinomial probability\n",
    "print(\"Multinomial Probability\")\n",
    "print(log_multinomial_prob(training_genome, testing_genome))\n",
    "print()\n",
    "\n",
    "# Calculate the log of the order 3 Markov probability\n",
    "print(\"Order 3 Markov Probability\")\n",
    "print(log_markov_prob(training_genome, testing_genome))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7. Generate a random sequence of 20,000 characters using a Markov model \n",
    "# of order 3 trained on the human mitochondrial genome. \n",
    "# See random.fasta for the genome\n",
    "\n",
    "import random\n",
    "\n",
    "def generate_random_sequence(training_genome, length):\n",
    "    order = 3\n",
    "    \n",
    "    # Calculate the nucleotide frequencies from training\n",
    "    nucleotide_frequencies = freq_nucleotides(training_genome)\n",
    "    \n",
    "    # Create dictionary of the transitions for all possible 4 letter combinations\n",
    "    transitions = {}\n",
    "    for nucleotide1 in nucleotides:\n",
    "        for nucleotide2 in nucleotides:\n",
    "            for nucleotide3 in nucleotides:\n",
    "                t_sum = 0.0\n",
    "                tmp = {}\n",
    "                for nucleotide4 in nucleotides:\n",
    "                    seq = nucleotide1 + nucleotide2 + nucleotide3 + nucleotide4\n",
    "                    # Find all occurences of this 4 nucleotide sequence\n",
    "                    count = len(re.findall(seq, training_genome, overlapped=True))\n",
    "                    tmp[seq] = count\n",
    "            \n",
    "                    t_sum += count\n",
    "                \n",
    "                # Find transition value by dividing by sum of transitions for this final letter\n",
    "                for key in tmp:\n",
    "                    transitions[key] = tmp[key] / t_sum\n",
    "                    \n",
    "    random_genome = \"\"\n",
    "    \n",
    "    border = {}\n",
    "    border['A'] = nucleotide_frequencies['A']\n",
    "    border['T'] = border['A'] + nucleotide_frequencies['T']\n",
    "    border['C'] = border['T'] + nucleotide_frequencies['C']\n",
    "    border['G'] = border['C'] + nucleotide_frequencies['G']\n",
    "    \n",
    "    for i in range(0, order):\n",
    "        rand = random.random()\n",
    "        if rand <= border['A']:\n",
    "            random_genome += 'A'\n",
    "        elif rand > border['A'] and rand < border['T']:\n",
    "            random_genome += 'T'\n",
    "        elif rand > border['T'] and rand < border['C']:\n",
    "            random_genome += 'C'\n",
    "        else:\n",
    "            random_genome += 'G'\n",
    "            \n",
    "    for i in range(0, length-order):\n",
    "        current_seq = random_genome[i:i+order]\n",
    "        border = {}\n",
    "        border['A'] = transitions[current_seq + 'A']\n",
    "        border['T'] = border['A'] + transitions[current_seq + 'T']\n",
    "        border['C'] = border['T'] + transitions[current_seq + 'C']\n",
    "        border['G'] = border['C'] + transitions[current_seq + 'G']\n",
    "        \n",
    "        rand = random.random()\n",
    "        \n",
    "        if rand <= border['A']:\n",
    "            random_genome += 'A'\n",
    "        elif rand > border['A'] and rand < border['T']:\n",
    "            random_genome += 'T'\n",
    "        elif rand > border['T'] and rand < border['C']:\n",
    "            random_genome += 'C'\n",
    "        else:\n",
    "            random_genome += 'G'\n",
    "        \n",
    "    return random_genome\n",
    "\n",
    "# Input the human mitochondrial genome for training\n",
    "training_header, training_genome = input_genome(\"human_mito.fasta\", 1)\n",
    "\n",
    "# Generate a random 20,000 nucleotide sequence\n",
    "random_genome = generate_random_sequence(training_genome, 20000)\n",
    "\n",
    "# Output the randome sequence genome\n",
    "output_genome(\"random.fasta\", \"> randomly generated using order 3 Markov model trained on the human mitochondrial genome\", random_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
