{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework 2\n",
    "### COSC 594 Bioinformatics\n",
    "### Started 01/29/18, Due 02/06/18 \n",
    "### Elliot Greenlee\n",
    "\n",
    "1\\. See included files, fruit_fly.fasta and human.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input a genome from a file in fasta format\n",
    "def input_genome(filename, chosen_gen):    \n",
    "    with open(filename, 'r') as f:\n",
    "        header = f.readline()\n",
    "        current_gen = 1\n",
    "        genome = \"\"\n",
    "        for line in f:\n",
    "            if line[0] is '>':\n",
    "                if current_gen is chosen_gen:\n",
    "                    genome = genome.replace('\\n', '')\n",
    "                    genome = genome.replace('N', 'A')\n",
    "                    return header, genome\n",
    "                else:\n",
    "                    header = line\n",
    "                    current_gen += 1\n",
    "            else:\n",
    "                genome = genome + line\n",
    "          \n",
    "        if current_gen is chosen_gen:\n",
    "            genome = genome.replace('\\n', '')\n",
    "            genome = genome.replace('N', 'A')\n",
    "            return header, genome\n",
    "        else:\n",
    "            print(\"This file only has {} genomes\".format(current_gen))\n",
    "            exit(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: -1016\n",
      "\n",
      "fruit fly TTCGCACGGCGTGCGTTTGGCTGAACACAGCAGTCTCTTGGCTAAAGCTTTCATGAGCAGTGCATGTAATAAAAACTGAGATCCAACTATGTTTACATTG\n",
      "human     AT-GCA-G---------------AACA--G---TCAC------A--GC------G-G-AGTG-A---A-T-----C--AG--C----T-------C---G\n",
      "\n",
      "fruit fly CAACCAACTCCAACTGCTATAGGCACCGTGGTTCCCCCATGGTCAGCGGGAACATTGATAGAGCGCCTGCCGTCTTTAGAAGACATGGCTCACAAGGGTC\n",
      "human     ---------------G-T---GGT---GTC-TT------TG-TCAACGGG--C---G---GC-CAC-TGCCG------GA---C-T--C-CAC------C\n",
      "\n",
      "fruit fly ACAGTGGAGTAAATCAGCTGGGTGGCGTTTTTGTTGGAGGAAGGCCTTTGCCAGATTCAACACGGCAAAAAATTGTCGAACTGGCACATTCTGGAGCTCG\n",
      "human     -C-G-GCAG-AAG--A--T---TG---T---------AG-A-G-C-TA-GC----T-CA-CA-G-C--------G--G----GGC-C---C-GG--C-CG\n",
      "\n",
      "fruit fly GCCATGTGATATTTCTCGAATTCTGCAAGTATCAAATGGATGTGTGAGCAAAATTCTCGGG-AGGTATTATGAAACAGGAAGCATACGACCACGTGCTAT\n",
      "human     ----TGCGACATTTCCCGAATTCTGCAGGTGTCCAACGGATGTGTGAGTAAAATTCT-GGGCAGGTATTACGAGACTGG---C-T-C--C-A--T-C-A-\n",
      "\n",
      "fruit fly CGGAGGATCCAAGCCACGTGTGGCCACAGCCGAAGTCGTTAGCAAAATTTCGCAGTACAAACGCGAGTGTCCTAGCATATTTGCTTGGGAAATTCGGGAT\n",
      "human     -G-A----CC---C-A-G-G-G-C-A-ATC-G--GT-G---G-----T-----AGTA-AA-C-CGAG------AG--TA---GC--G--A----C----T\n",
      "\n",
      "fruit fly AGATTACTTCAGGAGAACGTTTGTACTAACGATAATATACCAAGTGTGTCCTCAATAAACCGTGTATTGAGAAACTTGGCTGCGCAAAAGGAGCAGCAAA\n",
      "human     ------C--CAG-A-A--GTT-GTA--AGC-A-AA-ATA----G-----CC-CA-------GT--AT--A-A-----G-C-G-G-----G-AGT-GC---\n",
      "\n",
      "fruit fly GCACGGGATCCGGGAGCTCCAGCACATCCGCCGGCAACTCAATCAGCGCAAAAGTGTCTGTCAGCATCGGTGGCAACGTGAGCAATGTGGCAAGCGGATC\n",
      "human     -C-CG---TCC---ATCTTT-GCT--T--G--GG-AA---A-TC--CG-AGA-----C----AG-AT---T---A-C-TG-----T----C---CG-A--\n",
      "\n",
      "fruit fly GAGAGGCACGTTGAGCTCTTCCACCGATCTTATGCAGACAGCCACTCCTCTTAACTCTTCGGAAAGCGGTGGCGCAACGAACTCCGGGGAGGGTAGTGAA\n",
      "human     G-G-GG---GT----CTGTACCAACGAT---A---A--CA-----T------A-C-C-----AA-GCG-TGTC--ATC-AA-T------A----A---A-\n",
      "\n",
      "fruit fly CAGGAGG-CGATTTACGAGAAGCTTCGGCT-GTTAAATACTCAGCACGCTGCAGGACCAGGACCACTGGAGCCTGCCAGAGCAGCGCCCTTGGTAGGTCA\n",
      "human     CAG-AGTTC--TT--CGC-AACCT--GGCTAGCGAAA-A----GCA------A---C-AG-A----TGG-GC--GC-AGA-C-G-GCA-T-G-TATG--A\n",
      "\n",
      "fruit fly ATCACCCAACCACCTAGGAACCCGATCCAGCCACCCCCAGCTGGTGCACGGTAACCATCAGGCACTACAGCAGCATCAACAGCAGAGCTGGCCGCCCCGT\n",
      "human     -T-A---AACTA---AGGAT---GTT---G--A-----A-C-GG-GCA-G--A-CC----GG-A--A--GC----T-----G--G-G--G-CA-CCC-GC\n",
      "\n",
      "fruit fly CACTATTCCGGATCTTGGTACCCCACCTCTCTTAGCGAAATACCCATCTCATCGGCTCCCAATATCGCATCCGTTACGGCGTATGCATCAGGACCTTCAC\n",
      "human     C-CT-----GG-T-T-GGTA-------TC-CGG-G-GA----C---T-TC---GG-T---------GC--C----A-GG-G----CA--A---CCT--AC\n",
      "\n",
      "fruit fly TTGCTCACTCACTGAGTCCACCCAACGACATCAAAAGCCTGGCCAGTATCGGTCACCAGAGAAACTGCCCCGTTGCAACGGAGGACATACATTTAAAAAA\n",
      "human     --GC--A---A--GA-T---------G---------GC-TG-CCAGCA------AC-AG-GAA---G----GA-G----GG-GGA---------------\n",
      "\n",
      "fruit fly AGAACTTGATGGTCATCAGTCCGATGAAACGGGCTCCGGTGAAGGTGAAAACTCCAATGGTGGCGCTTCAAATATAGGAAACACTGAGGATGATCAAGCT\n",
      "human     -GA----GA-----AT-A--CC-A--A--C----TCCA-TCA-G-T------TCCAACGGAGAAG----A--T-T------CA--GA---TGA----G--\n",
      "\n",
      "fruit fly CGGCTCATACTAAAAAGAAAGTTGCAACGCAATCGAACATCTTTCACGAACGACCAGATAGACAGTCTTGAAAAAGAGTTTGAACGAACACACTATCCAG\n",
      "human     -G-CTCA-A--A---------T-GCGACT---TC-AGC-T-------GAA-G-C--G---GA-AG-CT-GCAAA-GAAAT--A--GAACA---T--CC--\n",
      "\n",
      "fruit fly ATGTTTTTGCCCGCGAACGTTTGGCTGGAAAGATTGGGTTGCCAGAGGCAAGAATTCAGGTTTGGTTCTCAAACCGTCGAGCAAAATGGCGTCGCGAGGA\n",
      "human     -T-TTA---CCC---AA-GA--G-C---AAA--TTGAG--GCCCT-GG-A-GAA---AG-----------A----GT---------T----T-G--AG-A\n",
      "\n",
      "fruit fly GAAGCTGCGAAACCAGCGAAGAACACCAAATTCCACAGGAGCTAGTGCAACTTCTTCCTCTACATCGGCAACCGCCTCTTTGACTGACAGCCCTAACAGC\n",
      "human     GAA-C--C----C-ATT-A-----------T-CCA--G-A--T-GTG----TT-TGCC-CGAGA-----AA--GACT-----A--G-CAGCCA-AA-A--\n",
      "\n",
      "fruit fly CTAAGTGCTTGTTCCTCGCTGCTGTCCGGATCAGCTGGGGGTCCCTCAGTCAGTACCATTAATGGCTTATCGTCTCCAAGCACATTGTCTACTAATGTCA\n",
      "human     -TA-GA--T----C-TA-C--CTG-----A--AGCAAG---------A---A-TAC-A-----GG-T-AT-G-------G----TT-T-T-CTAAT--C-\n",
      "\n",
      "fruit fly ATGCTCCAACGCTTGGCGCTGGGATCGATAGCTCTGAAAGCCCAACACCAATCCCGCACATTCGGCCTAGCTGCACCTCTGACAATGACAATGGTCGTCA\n",
      "human     --G----AA-G---GGC-CA---A---AT-G----GA--G---AAGA--------G-A-A---G----A-----A------A-AA---C--TG------A\n",
      "\n",
      "fruit fly AAGTGAAGATTGCAGAAGAGTTTGTTCTCCATGCCCACTTGGCGTTGGCGGGCATCAAAATACTCATCATATCCAGAGCAATGGTCACGCCCAAGGTCAT\n",
      "human     --G-GAA--T--CAGA-GAA---G------A--C--A---GGC-----CAG-CA--A-----C--A-CA---CC-------T----A-G-------TCAT\n",
      "\n",
      "fruit fly GCACTTGTTCCTGCCATTTCGCCACGACTCAATTTTAATAGTGGTAGTTTCGGCGCGATGTACTCCAACATGCATCATACGGCGTTATCCATGAGCGATT\n",
      "human     --A-TT---CCT---AT--C---A-G-C--A---------GT---AGTTTCAGC---A----C-C-AG--TG--TC-TAC--CA--A-CCA--A----TT\n",
      "\n",
      "fruit fly CATATGGGGCGGTTACGCCGATTCCGAGCTTTAACCACTCAGCTGTCGGTCCGCTGGCTCCGCCATCGCCAATACCGCAACAGGGCGATCTTACCCCTTC\n",
      "human     CC-A-----C----A-----A--CC---C---A-CCAC--A-C---CGGT----T---TCCTCCTTCAC-A-T-CTG-----G--C--TCC-A----T--\n",
      "\n",
      "fruit fly CTCGTTATATCCGTGCCACATGACCCTACGACCCCCTCCGATGGCTCCCGCTCACCATCACATCGTGCCGGGTGACGGTGGCAGACCTGCGGGCGTTGGC\n",
      "human     ---GTTG-----G-GCCG-A--AC---A-GACAC------A-G-C-CC---TCACAA--ACA-C---C----T-ACAGCG-CT--C-TGC---CG----C\n",
      "\n",
      "fruit fly CTAGGCAGTGGCCAATCTGCGAATTTGGGAGCAAGCTGCAGCGGATCGGGATACGAAGTGCTATCTGCCTACGCGTTGCCACCGCCCCCTATGGCGTCGA\n",
      "human     CTATGC-----CCA----GC---TTC---ACCATG--GCA----A-----ATA--A----C---CTGCCTA----T-GCAACC-CCC---A-G---TC--\n",
      "\n",
      "fruit fly GCTCTGCTGCTGATTCAAGCTTCTCAGCCGCGTCCAGTGCCAGCGCTAATGTGACCCCACATCACACCATAGCCCAAGAATCATGCCCCTCTCCGTGTTC\n",
      "human     -C-C--CAGC-----CA-G------A-CC---TCC--T-C-A----TA------C------TC-C----T-GC--ATG---C-TGCCC------------\n",
      "\n",
      "fruit fly AAGCGCGAGCCACTTTGGAGTTGCTCACAGTTCTGGGTTTTCGTCCGACCCGATTTCACCGGCTGTATCTTCGTATGCACATATGAGCTACAATTACGCG\n",
      "human     A--C-C-AGCC-CTTCGG--T-G---A-A-T---GGG----CG---GA---G-TT--A-----TG-AT------A--C-C-TA----C-AC-----C-C-\n",
      "\n",
      "fruit fly TCGTCCGCTAACACCATGACGCCTTCCTCCGCCAGCGGCACATCAGCACACGTGGCCCCGGGAAAACAACAGTTCTTCGCCTCCTGTTTCTACTCACCGT\n",
      "human     -C--CC-C-A-CA---T-A----T------GC-AG----ACA-CA-CA----TG--------AA--CA---GT-CA--GCC----------A---A---T\n",
      "\n",
      "fruit fly GGGTCTAGGAACAGACTGGCGATTTGAGCAGAGAAGCACTGCGAAAGGACTATTTACATAGTTGAATGTATATCTAAAGGAGGCCATAATAAATCGAATT\n",
      "human     GGG-C-A----C---CT--CG----G-GCA------C-C----A-----CT-T---CA-A-------------C-A--GGA--C--T-------C-A-TT\n",
      "\n",
      "fruit fly TACATATCTCTTGAAAAATAATGGAGGTTGTAGAAAAATACATTTGTATGTATAAATTATATAGTTCCGCCCATTAAATCCAATCTATAGTGTAGAATAA\n",
      "human     TCC----C-CT-G---------G----T-GT-G-----T-CAGTTCCA-GT-TCAA-------GTTCC-CGGA--AG-TG-AACCT---G---A---TA-\n",
      "\n",
      "fruit fly TTGGTGTAAATTAAATGATATAATTTTGACAAATAAAAAAAAAAAAAAAAA\n",
      "human     T-G-TCTCAAT-AC-TGGCC-AAGATT-ACAG-TAA---------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 2. This program, globalign, computes the global alignment score of the human and fruit fly sequences. \n",
    "# Current parameters are match: +2, mismatch: -1, and gap: -2. \n",
    "# This program only calculates one of the possible best tracebacks, prioritizing left, diagonal, then up\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def globalign(sequence1, sequence2, match, mismatch, gap):\n",
    "    # Initialize tables\n",
    "    scores = np.zeros((len(sequence1) + 1, len(sequence2) + 1), dtype=int)\n",
    "    traceback = np.zeros((len(sequence1) + 1, len(sequence2) + 1), dtype='i,i')\n",
    "    \n",
    "    # Add starting scores and traceback\n",
    "    traceback[0][0] = (-1, -1)\n",
    "\n",
    "    # First row\n",
    "    for c in range(len(scores[0])):\n",
    "        scores[0][c] = gap * c\n",
    "        \n",
    "        if c is not 0:\n",
    "            traceback[0][c] = (0, c-1)\n",
    "    \n",
    "    # First column\n",
    "    for r in range(len(scores)):\n",
    "        scores[r][0] = gap * r\n",
    "        \n",
    "        if r is not 0:\n",
    "            traceback[r][0] = (r-1, 0)\n",
    "        \n",
    "    # Calculate scores and traceback, row by row\n",
    "    for r in range(1, len(scores)):\n",
    "        for c in range(1, len(scores[0])):\n",
    "            # Calculate the three possible scores\n",
    "            gap2sequence1 = scores[r-1][c] + gap # Look up\n",
    "            gap2sequence2 = scores[r][c-1] + gap # Look left\n",
    "            \n",
    "            if sequence1[r-1] is sequence2[c-1]:  # Look diagonal\n",
    "                sequence2sequence = scores[r-1][c-1] + match\n",
    "            else:\n",
    "                sequence2sequence = scores[r-1][c-1] + mismatch\n",
    "                \n",
    "            # Compare scores to find the largest and the correct traceback(s)\n",
    "            if gap2sequence2 >= gap2sequence1 and gap2sequence2 >= sequence2sequence:\n",
    "                # left\n",
    "                scores[r][c] = gap2sequence2\n",
    "                traceback[r][c] = (r, c-1)\n",
    "            elif gap2sequence1 >= gap2sequence2 and gap2sequence1 >= sequence2sequence:\n",
    "                # up\n",
    "                scores[r][c] = gap2sequence1\n",
    "                traceback[r][c] = (r-1, c)\n",
    "            elif sequence2sequence >= gap2sequence1 and sequence2sequence >= gap2sequence2:\n",
    "                # diagonal\n",
    "                scores[r][c] = sequence2sequence\n",
    "                traceback[r][c] = (r-1, c-1)\n",
    "    \n",
    "    r = len(sequence1)\n",
    "    c = len(sequence2)\n",
    "    trace = traceback[r][c]\n",
    "    rev_seq1 = \"\"\n",
    "    rev_seq2 = \"\"\n",
    "    \n",
    "    while r != 0 or c != 0:\n",
    "        r_diff = trace[0] - r\n",
    "        c_diff = trace[1] - c\n",
    "        \n",
    "        if r_diff == -1 and c_diff == -1:\n",
    "            # diagonal trace\n",
    "            rev_seq1 += \"{}\".format(sequence1[r-1])\n",
    "            rev_seq2 += \"{}\".format(sequence2[c-1])\n",
    "            r += -1\n",
    "            c += -1\n",
    "        elif r_diff == -1:\n",
    "            # up trace\n",
    "            rev_seq1 += \"{}\".format(sequence1[r-1])\n",
    "            rev_seq2 += \"-\"\n",
    "            r += -1\n",
    "        elif c_diff == -1:\n",
    "            # left trace\n",
    "            rev_seq1 += \"-\"\n",
    "            rev_seq2 += \"{}\".format(sequence2[c-1])\n",
    "            c += -1\n",
    "            \n",
    "        trace = traceback[r][c]\n",
    "    \n",
    "    seq1_align = rev_seq1[::-1]\n",
    "    seq2_align = rev_seq2[::-1]\n",
    "    align_score = scores[-1][-1]\n",
    "        \n",
    "    return align_score, seq1_align, seq2_align\n",
    "    \n",
    "    \n",
    "# Constants\n",
    "match = 2\n",
    "mismatch = -1\n",
    "gap = -2\n",
    "\n",
    "# Read in the sequences\n",
    "header, fruit_fly = input_genome(\"fruit_fly.fasta\", 1)\n",
    "header, human = input_genome(\"human.fasta\", 1)\n",
    "\n",
    "align_score, fruit_fly_align, human_align = globalign(fruit_fly, human, match, mismatch, gap)\n",
    "\n",
    "print(\"Score: {}\".format(align_score))\n",
    "print()\n",
    "\n",
    "for i in range(0, len(fruit_fly_align), 100):\n",
    "    print(\"fruit fly \" + fruit_fly_align[i:i+100])\n",
    "    print(\"human     \" + human_align[i:i+100])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 820\n",
      "\n",
      "fruit fly ATGGCTCACAAGGGTCACAGTGGAGTAAATCAGCTGGGTGGCGTTTTTGTTGGAGGAAGGCCTTTGCCAGATTCAACACGGCAAAAAATTGTCGAACT-G\n",
      "human     ATG-C--AGAACAGTCACAGCGGAGTGAATCAGCTCGGTGGTGTCTTTGTCAACGGGCGGCCACTGCCGGACTCCACCCGGCAGAAGATTGTAGAGCTAG\n",
      "\n",
      "fruit fly -GCACATTCTGGAGCTCGGCCATGTGATATTTCTCGAATTCTGCAAGTATCAAATGGATGTGTGAGCAAAATTCTCGGGAGGTATTATGAAACAGGAAGC\n",
      "human     CTCACAG-CGGG-GCCCGGCCGTGCGACATTTCCCGAATTCTGCAGGTGTCCAACGGATGTGTGAGTAAAATTCTGGGCAGGTATTACGAGACTGGCTCC\n",
      "\n",
      "fruit fly AT-ACGA-CCACGTGCTATCGGAGG-A-TCCAAGCCACGTGTGGCCACAGCCGAAGTCGTTAGCAAAATTTCGCAGTACAAACGCGAGTGTCC-TAGCAT\n",
      "human     ATCA-GACCCAGG-GCAATCGGTGGTAGTA-AACCGA-GAGTAGCGACTCCAGAAGTTGTAAGCAAAATAGCCCAGTATAAGCGGGAGTGCCCGTC-CAT\n",
      "\n",
      "fruit fly ATTTGCTTGGGAAATTCGGGATAGATTACT-TCAGGAGAACGTTTGTACTAACGATAATATACCAAGTGTGTCCTCAATAAACCGTG-TATTGAGAAACT\n",
      "human     CTTTGCTTGGGAAATCCGAGACAGATTACTGTCCG-AGGGGGTCTGTACCAACGATAACATACCAAGCGTGTCATCAATAAACAGAGTTCTTC-GCAACC\n",
      "\n",
      "fruit fly TGGCT-GCGCAAAAGGAGCAGCAAAGCACGGGATCCGGGA-GCTCCAGCACATCCGCCGGCAACTCAATCAGCGCAAAAGTGTCTGTCAGCATCGGTGGC\n",
      "human     TGGCTAGCG-AAAAGCAACAGATGGGCGCAG-A-C-GGCATG-TAT-G-ATAAACTAAGGATGTTGAA-CGG-GCAGACC-G---GA-AGC-T-GG-GGC\n",
      "\n",
      "fruit fly AACGTGAGCAATGTGGCAAGCGGAT-CGAGAGGCACGTTGAGCTCTTCCACCG-ATCTTATGC-AGACAGCCACTCCTCTTAACTCTTCGGAAAGCGGTG\n",
      "human     ACCC-GCCC--TG-GTT--G-GTATCCG-G-GG-ACTTCG-G-TG--CCAGGGCAACCTACGCAAGATGGCTGC-CAGC--AACAGGAAGGAGGG-GGAG\n",
      "\n",
      "fruit fly GCGCA-ACGAACTCCGGGGAGGGTAGTGAACAGGAGGCGATTTACGA-GAAGCTTCGGCTGTTAAAT-A-CT-CAGC--ACGCTGC--AGGACCAGGAC-\n",
      "human     A-GAATACCAACTCCATC-AGT-TCC--AAC-GGAGAAGATTCA-GATGAGGCT-CAAATGCGACTTCAGCTGAAGCGGAAGCTGCAAAGAAATAGAACA\n",
      "\n",
      "fruit fly -CACTGGAGCC-TG-CCAGA--GCAGCGCCCTTG-G-TAG-GTCAATCACCCAACCACCTAGGAACCCGATCCAGCCACCCCCAG-CTG-GT-GCACGGT\n",
      "human     TC-CTTTACCCAAGAGCAAATTG-AG-GCCCTGGAGAAAGAGTT--TGAGAGAACC-CATT--ATCCAGATGT-GTTTGCCCGAGAAAGACTAGCA-GCC\n",
      "\n",
      "fruit fly AACCATCAGGCACTA-CAGCAGCATCAACAGCAGAGCTGGCCGCCCCGTC-ACTATTCCGGATCTTG--GTACCCCACCTCTCTTAGCGAAATACCCATC\n",
      "human     AAA-AT-AGAT-CTACCTGAAGCAAGAATA-CAGGTATGGTTTTCTAATCGAAGGG-CCAAATGGAGAAG-AGAAGAAAAA-CTGAG-GAATCAGAGAAG\n",
      "\n",
      "fruit fly TCATCGGCTCCCAATATCGCATCCGT--TACGGCGTATGCATCAGGACCTTCACTTGCTCACTCACTGAGT-CCACCCAA--CGACATCAAAAGCCTGGC\n",
      "human     ACA--GGCCAGCAACA-CACCTA-GTCATATTCC-TAT-CAGCAGTAGTTTCA---GC--AC-CAGTGTCTACCAACCAATTCCACAACCCA--CCACAC\n",
      "\n",
      "fruit fly CAGTATCGGTCACCAGAGAAACT-GC-CC-CGTTGCAACGGA-GGACATA-CATTTA-AAAAAAGAACTTGATGGTCATCAGTCCGATGAAACGGGCTCC\n",
      "human     CGGTTTCC-TC-CTTCACAT-CTGGCTCCATGTTGGGCCGAACAGACACAGCCCTCACAAACACCTACA-GC-GCTCTGCCG-CCTATGCC-CAG-CTTC\n",
      "\n",
      "fruit fly GGTGAAGGTGAA-AA-CT-CCAATGGTGGCGCTTCA-AATATAG-GAAACACTGAGGATGATCAAGC-T-CGGCTCA-TA---CTAAAAAGAAAGTTGC-\n",
      "human     ACC-ATGGCAAATAACCTGCCTATGCAACCCC--CAGTCCCCAGCCAGAC-CTCCTCATACTCCTGCATGCTGCCCACCAGCCCTTCGGTGAATGG-GCG\n",
      "\n",
      "fruit fly AACGCA--AT-CGAACA-TCTTTCAC-GA-AC-GAC-CAGAT-AGACAGTCTTGAAAAAGAGTTTGAACGAACA-CACTATC--CA-GA-TGTTTTTGCC\n",
      "human     GAGTTATGATACCTACACCCCCCCACATATGCAGACACACATGA-ACAGTCA-GCCAATGGGCACCT-CGGGCACCACT-TCAACAGGACTCATTTC-CC\n",
      "\n",
      "fruit fly CGCGAACGTTTGGCTG-GAAAGATTGGGTTGCCAGAGGCAAGAATTCAGGTTTGGTTCTCAA-ACCGTCGAGCAAAATGGC-GT\n",
      "human     CT-G---GTGTGTCAGTTCCAGTTCAAGTTCCCGGAAGT--GAAC-CTGATATG-T-CTCAATACTGGC---CAAGATTACAGT\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 3. This program, localign, computes the local alignment score of the human and fruit fly sequences. \n",
    "# Current parameters are match: +2, mismatch: -1, and gap: -2. \n",
    "# This program only calculates one of the possible best tracebacks, prioritizing left, diagonal, then up\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def localign(sequence1, sequence2, match, mismatch, gap):\n",
    "    # Initialize tables\n",
    "    scores = np.zeros((len(sequence1) + 1, len(sequence2) + 1), dtype=int)\n",
    "    traceback = np.zeros((len(sequence1) + 1, len(sequence2) + 1), dtype='i,i')\n",
    "    \n",
    "    # Add starting scores and traceback\n",
    "    # First row\n",
    "    for c in range(len(scores[0])):\n",
    "        scores[0][c] = 0\n",
    "        traceback[0][c] = (-1, -1)\n",
    "    \n",
    "    # First column\n",
    "    for r in range(len(scores)):\n",
    "        scores[r][0] = 0\n",
    "        traceback[r][0] = (-1, -1)\n",
    "            \n",
    "    best_r = 0\n",
    "    best_c = 0\n",
    "    best_score = 0\n",
    "        \n",
    "    # Calculate scores and traceback, row by row\n",
    "    for r in range(1, len(scores)):\n",
    "        for c in range(1, len(scores[0])):\n",
    "            # Calculate the three possible scores\n",
    "            gap2sequence1 = scores[r-1][c] + gap # Look up\n",
    "            gap2sequence2 = scores[r][c-1] + gap # Look left\n",
    "            \n",
    "            if sequence1[r-1] is sequence2[c-1]:  # Look diagonal\n",
    "                sequence2sequence = scores[r-1][c-1] + match\n",
    "            else:\n",
    "                sequence2sequence = scores[r-1][c-1] + mismatch\n",
    "                \n",
    "            # Compare scores to find the largest and the correct traceback(s)\n",
    "            if gap2sequence1 >= gap2sequence2 and gap2sequence1 >= sequence2sequence:\n",
    "                # up\n",
    "                scores[r][c] = gap2sequence1\n",
    "                traceback[r][c] = (r-1, c)\n",
    "            elif sequence2sequence >= gap2sequence1 and sequence2sequence >= gap2sequence2:\n",
    "                # diagonal\n",
    "                scores[r][c] = sequence2sequence\n",
    "                traceback[r][c] = (r-1, c-1)\n",
    "            elif gap2sequence2 >= gap2sequence1 and gap2sequence2 >= sequence2sequence:\n",
    "                # left\n",
    "                scores[r][c] = gap2sequence2\n",
    "                traceback[r][c] = (r, c-1)\n",
    "                \n",
    "            if scores[r][c] < 0:\n",
    "                scores[r][c] = 0\n",
    "                traceback[r][c] = (-1, -1)\n",
    "                \n",
    "            if scores[r][c] >= best_score:\n",
    "                best_score = scores[r][c]\n",
    "                best_r = r\n",
    "                best_c = c\n",
    "    \n",
    "    r = best_r\n",
    "    c = best_c\n",
    "    trace = traceback[r][c]\n",
    "    rev_seq1 = \"\"\n",
    "    rev_seq2 = \"\"\n",
    "    \n",
    "    while trace[0] != -1 or trace[1] != -1:\n",
    "        r_diff = trace[0] - r\n",
    "        c_diff = trace[1] - c\n",
    "        \n",
    "        if r_diff == -1 and c_diff == -1:\n",
    "            # diagonal trace\n",
    "            rev_seq1 += \"{}\".format(sequence1[r-1])\n",
    "            rev_seq2 += \"{}\".format(sequence2[c-1])\n",
    "            r += -1\n",
    "            c += -1\n",
    "        elif r_diff == -1:\n",
    "            # up trace\n",
    "            rev_seq1 += \"{}\".format(sequence1[r-1])\n",
    "            rev_seq2 += \"-\"\n",
    "            r += -1\n",
    "        elif c_diff == -1:\n",
    "            # left trace\n",
    "            rev_seq1 += \"-\"\n",
    "            rev_seq2 += \"{}\".format(sequence2[c-1])\n",
    "            c += -1\n",
    "            \n",
    "        trace = traceback[r][c]\n",
    "    \n",
    "    seq1_align = rev_seq1[::-1]\n",
    "    seq2_align = rev_seq2[::-1]\n",
    "    align_score = best_score\n",
    "        \n",
    "    return align_score, seq1_align, seq2_align\n",
    "    \n",
    "    \n",
    "# Constants\n",
    "match = 2\n",
    "mismatch = -1\n",
    "gap = -2\n",
    "\n",
    "# Read in the sequences\n",
    "header, fruit_fly = input_genome(\"fruit_fly.fasta\", 1)\n",
    "header, human = input_genome(\"human.fasta\", 1)\n",
    "\n",
    "align_score, fruit_fly_align, human_align = localign(fruit_fly, human, match, mismatch, gap)\n",
    "\n",
    "print(\"Score: {}\".format(align_score))\n",
    "print()\n",
    "\n",
    "for i in range(0, len(fruit_fly_align), 100):\n",
    "    print(\"fruit fly \" + fruit_fly_align[i:i+100])\n",
    "    print(\"human     \" + human_align[i:i+100])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3\\. Discussion. After viewing the global alignment, it did not seem these two sequences were related. However, the local alignment seems to make it clear that a large portion of the two sequences are very similar. The matching portion for local alignment runs from 184 to 1489 in the fruit fly sequence, and from 0 to 1267 in the human sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 620\n",
      "\n",
      "human       CCAAGTATTGACTCACCCATCAACAACCGCTATGTATTTCGTACATTACTGCCAGCCACCATGAATATTGTACGGTACCATAAATACTTGACCACCTGTA\n",
      "neanderthal CCAAGTATTGACTCACCCATCAACAACCGCCATGTATTTCGTACATTACTGCCAGCCACCATGAATATTGTACAGTACCATAATTACTTGACTACCTGTA\n",
      "\n",
      "human       GTACATAAAAACCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCACACATCAACTGCAACTCCAAAG\n",
      "neanderthal ATACATAAAAACCTAATCCACATCAACCCCCCCCCCCCATGCTTACAAGCAAGCACAGCAATCAACCTTCAACTGTCATACATCAACTACAACTCCAAAG\n",
      "\n",
      "human       CCACCCCT-CACCCACTAGGATACCAACAAACCTACCCACCCTTAACAGTACATAGTACATAAAGCCATTTACCGTACATAGCACATTACAGTCAAATCC\n",
      "neanderthal ACACCCTTACACCCACTAGGATATCAACAAACCTACCCACCCTTGACAGTACATAGCACATAAAGTCATTTACCGTACATAGCACATTATAGTCAAATCC\n",
      "\n",
      "human       CTTCTCGTCCCCATGGATGACCCCCCTCAGATAGGGGTCCCTTGA\n",
      "neanderthal CTTCTCGCCCCCATGGATGACCCCCCTCAGATAGGGGTCCCTTGA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 4. See included files, human_mito.fasta and neander_sample.fasta\n",
    "# This program, egfalign, computes the end-gap free alignment score of the human mitochondrial and neanderthal sequences. \n",
    "# Current parameters are match: +2, mismatch: -1, and gap: -2. \n",
    "# This program only calculates one of the possible best tracebacks, prioritizing left, diagonal, then up\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def egfalign(sequence1, sequence2, match, mismatch, gap):\n",
    "    # Initialize tables\n",
    "    scores = np.zeros((len(sequence1) + 1, len(sequence2) + 1), dtype=int)\n",
    "    traceback = np.zeros((len(sequence1) + 1, len(sequence2) + 1), dtype='i,i')\n",
    "    \n",
    "    # Add starting scores and traceback\n",
    "    # First row\n",
    "    for c in range(len(scores[0])):\n",
    "        scores[0][c] = 0 \n",
    "        traceback[0][c] = (-1, -1)\n",
    "    \n",
    "    # First column\n",
    "    for r in range(len(scores)):\n",
    "        scores[r][0] = 0\n",
    "        traceback[r][0] = (-1, -1)\n",
    "            \n",
    "    best_r = 0\n",
    "    best_c = 0\n",
    "    best_score = 0\n",
    "        \n",
    "    # Calculate scores and traceback, row by row\n",
    "    for r in range(1, len(scores)):\n",
    "        for c in range(1, len(scores[0])):\n",
    "            # Calculate the three possible scores\n",
    "            gap2sequence1 = scores[r-1][c] + gap # Look up\n",
    "            gap2sequence2 = scores[r][c-1] + gap # Look left\n",
    "            \n",
    "            if sequence1[r-1] is sequence2[c-1]:  # Look diagonal\n",
    "                sequence2sequence = scores[r-1][c-1] + match\n",
    "            else:\n",
    "                sequence2sequence = scores[r-1][c-1] + mismatch\n",
    "                \n",
    "            # Compare scores to find the largest and the correct traceback(s)\n",
    "            if gap2sequence1 >= gap2sequence2 and gap2sequence1 >= sequence2sequence:\n",
    "                # up\n",
    "                scores[r][c] = gap2sequence1\n",
    "                traceback[r][c] = (r-1, c)\n",
    "            elif sequence2sequence >= gap2sequence1 and sequence2sequence >= gap2sequence2:\n",
    "                # diagonal\n",
    "                scores[r][c] = sequence2sequence\n",
    "                traceback[r][c] = (r-1, c-1)\n",
    "            elif gap2sequence2 >= gap2sequence1 and gap2sequence2 >= sequence2sequence:\n",
    "                # left\n",
    "                scores[r][c] = gap2sequence2\n",
    "                traceback[r][c] = (r, c-1)\n",
    "                \n",
    "            # keep track of the best score in the last row and column\n",
    "            if ((r == len(scores)-1 ) or (c == len(scores[0])-1)) and scores[r][c] >= best_score:\n",
    "                best_score = scores[r][c]\n",
    "                best_r = r\n",
    "                best_c = c\n",
    "    \n",
    "    r = best_r\n",
    "    c = best_c\n",
    "    trace = traceback[r][c]\n",
    "    rev_seq1 = \"\"\n",
    "    rev_seq2 = \"\"\n",
    "    \n",
    "    while trace[0] != -1 or trace[1] != -1:\n",
    "        r_diff = trace[0] - r\n",
    "        c_diff = trace[1] - c\n",
    "        \n",
    "        if r_diff == -1 and c_diff == -1:\n",
    "            # diagonal trace\n",
    "            rev_seq1 += \"{}\".format(sequence1[r-1])\n",
    "            rev_seq2 += \"{}\".format(sequence2[c-1])\n",
    "            r += -1\n",
    "            c += -1\n",
    "        elif r_diff == -1:\n",
    "            # up trace\n",
    "            rev_seq1 += \"{}\".format(sequence1[r-1])\n",
    "            rev_seq2 += \"-\"\n",
    "            r += -1\n",
    "        elif c_diff == -1:\n",
    "            # left trace\n",
    "            rev_seq1 += \"-\"\n",
    "            rev_seq2 += \"{}\".format(sequence2[c-1])\n",
    "            c += -1\n",
    "            \n",
    "        trace = traceback[r][c]\n",
    "    \n",
    "    seq1_align = rev_seq1[::-1]\n",
    "    seq2_align = rev_seq2[::-1]\n",
    "    align_score = best_score\n",
    "        \n",
    "    return align_score, seq1_align, seq2_align\n",
    "    \n",
    "    \n",
    "# Constants\n",
    "match = 2\n",
    "mismatch = -1\n",
    "gap = -2\n",
    "\n",
    "# Read in the sequences\n",
    "header, human = input_genome(\"human_mito.fasta\", 1)\n",
    "header, neanderthal = input_genome(\"neander_sample.fasta\", 1)\n",
    "\n",
    "align_score, human_align, neanderthal_align = egfalign(human, neanderthal, match, mismatch, gap)\n",
    "\n",
    "print(\"Score: {}\".format(align_score))\n",
    "print()\n",
    "\n",
    "for i in range(0, len(human_align), 100):\n",
    "    print(\"human       \" + human_align[i:i+100])\n",
    "    print(\"neanderthal \" + neanderthal_align[i:i+100])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5\\. Discussion. Similarly to the local alignment calculated, the fruit fly alignment starts around 200 and ends around 1500, and the human alignment starts around 0, but ends much sooner around 800. This seems to be related to the blast algorithm choosing two separate local alignments rather than one single local alignment, and using a different scoring system. Despite the differences, it seems to confirm the local alignment results. The first of the two alignments is of length 397, identity 73%, and e-value 4e-65. The second of the two alignments is of length 203, identity 71%, and e-value 6e-25."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6\\. The \"mystery\" sequence has 9410 nucleotides, and a label of gi|224589819:c155604967-155595558 Homo sapiens chromosome 7, GRCh37.p9 Primary Assembly. \n",
    "\n",
    "a. blastn: The mystery sequence has 244 blast hits on 100 sequences. Three of these are an exact match to the query, with 100% cover and 100% identity. Homo sapiens sonic hedgehog (SHH), RefSeqGene on chromosome 7; Homo sapiens BAC clone RP11-69O3 from 7, complete sequence; and Human cosmid clone LL157A05 from 7q36, complete sequence. \n",
    "\n",
    "b. blastx: The mystery sequence has 252 blast hits on 100 sequences. The highest coverage is 15% by 8 sequences, and three of these have the highest identify of 99%. sonic hedgehog homolog (Drosophila) [Homo sapiens]; PREDICTED: sonic hedgehog protein [Pan troglodytes]; and PREDICTED: sonic hedgehog protein isoform X1 [Gorilla gorilla gorilla]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 620\n"
     ]
    }
   ],
   "source": [
    "# 7. Linear space # 4\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def egfalign_linear(sequence1, sequence2, match, mismatch, gap):\n",
    "    # Set sequence2 to be the shorter of the two sequences\n",
    "    if len(sequence2) > len(sequence1):\n",
    "        tmp = sequence2\n",
    "        sequence2 = sequence1\n",
    "        sequence1 = tmp\n",
    "    \n",
    "    # Initialize score rows\n",
    "    scores = np.zeros((2, len(sequence2) + 1), dtype=int)\n",
    "    \n",
    "    # Add starting scores\n",
    "    # First row\n",
    "    for c in range(len(scores[0])):\n",
    "        scores[0][c] = 0 \n",
    "            \n",
    "    best_score = 0\n",
    "        \n",
    "    # Calculate scores and traceback, row by row\n",
    "    for r in range(0, len(sequence1)):\n",
    "        # Figure out first and second by modding r\n",
    "        first = r % 2\n",
    "        second = (r + 1) % 2\n",
    "        \n",
    "        # set first column in the second row to 0\n",
    "        scores[second][0] = 0\n",
    "        \n",
    "        # change all rs to first or second\n",
    "        for c in range(1, len(sequence2) + 1):\n",
    "            # Calculate the three possible scores\n",
    "            gap2sequence1 = scores[first][c] + gap # Look up\n",
    "            gap2sequence2 = scores[second][c-1] + gap # Look left\n",
    "            \n",
    "            if sequence1[r-1] is sequence2[c-1]:  # Look diagonal\n",
    "                sequence2sequence = scores[first][c-1] + match\n",
    "            else:\n",
    "                sequence2sequence = scores[first][c-1] + mismatch\n",
    "                \n",
    "            # Compare scores to find the largest and the correct traceback(s)\n",
    "            if gap2sequence1 >= gap2sequence2 and gap2sequence1 >= sequence2sequence:\n",
    "                # up\n",
    "                scores[second][c] = gap2sequence1\n",
    "            elif sequence2sequence >= gap2sequence1 and sequence2sequence >= gap2sequence2:\n",
    "                # diagonal\n",
    "                scores[second][c] = sequence2sequence\n",
    "            elif gap2sequence2 >= gap2sequence1 and gap2sequence2 >= sequence2sequence:\n",
    "                # left\n",
    "                scores[second][c] = gap2sequence2\n",
    "                \n",
    "            # keep track of the best score in the last row and column\n",
    "            if ((r == len(sequence1)-1) or (c == len(sequence2))) and scores[second][c] >= best_score:\n",
    "                best_score = scores[second][c]\n",
    "                \n",
    "        # set first row = second\n",
    "        scores[first] = scores[second]\n",
    "    \n",
    "    align_score = best_score\n",
    "        \n",
    "    return align_score\n",
    "    \n",
    "    \n",
    "# Constants\n",
    "match = 2\n",
    "mismatch = -1\n",
    "gap = -2\n",
    "\n",
    "# Read in the sequences\n",
    "header, human = input_genome(\"human_mito.fasta\", 1)\n",
    "header, neanderthal = input_genome(\"neander_sample.fasta\", 1)\n",
    "\n",
    "align_score = egfalign_linear(human, neanderthal, match, mismatch, gap)\n",
    "\n",
    "print(\"Score: {}\".format(align_score))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
