{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1\\. These are downloaded in the file small.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input a genome from a file in fasta format\n",
    "def input_genome(filename, chosen_gen):    \n",
    "    with open(filename, 'r') as f:\n",
    "        header = f.readline()\n",
    "        current_gen = 1\n",
    "        genome = \"\"\n",
    "        for line in f:\n",
    "            if line[0] is '>':\n",
    "                if current_gen is chosen_gen:\n",
    "                    genome = genome.replace('\\n', '')\n",
    "                    genome = genome.replace('N', 'A')\n",
    "                    return header, genome\n",
    "                else:\n",
    "                    header = line\n",
    "                    current_gen += 1\n",
    "                    genome = \"\"\n",
    "            else:\n",
    "                genome = genome + line\n",
    "          \n",
    "        if current_gen is chosen_gen:\n",
    "            genome = genome.replace('\\n', '')\n",
    "            genome = genome.replace('N', 'A')\n",
    "            return header, genome\n",
    "        else:\n",
    "            print(\"This file only has {} genomes\".format(current_gen))\n",
    "            exit(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Compute the global alignment of two sequences\n",
    "def globalign(sequence1, sequence2, match, mismatch, gap):\n",
    "    # Initialize tables\n",
    "    scores = np.zeros((len(sequence1) + 1, len(sequence2) + 1), dtype=int)\n",
    "    traceback = np.zeros((len(sequence1) + 1, len(sequence2) + 1), dtype='i,i')\n",
    "    \n",
    "    # Add starting scores and traceback\n",
    "    traceback[0][0] = (-1, -1)\n",
    "\n",
    "    # First row\n",
    "    for c in range(len(scores[0])):\n",
    "        scores[0][c] = gap * c\n",
    "        \n",
    "        if c is not 0:\n",
    "            traceback[0][c] = (0, c-1)\n",
    "    \n",
    "    # First column\n",
    "    for r in range(len(scores)):\n",
    "        scores[r][0] = gap * r\n",
    "        \n",
    "        if r is not 0:\n",
    "            traceback[r][0] = (r-1, 0)\n",
    "        \n",
    "    # Calculate scores and traceback, row by row\n",
    "    for r in range(1, len(scores)):\n",
    "        for c in range(1, len(scores[0])):\n",
    "            # Calculate the three possible scores\n",
    "            gap2sequence1 = scores[r-1][c] + gap # Look up\n",
    "            gap2sequence2 = scores[r][c-1] + gap # Look left\n",
    "            \n",
    "            if sequence1[r-1] is sequence2[c-1]:  # Look diagonal\n",
    "                sequence2sequence = scores[r-1][c-1] + match\n",
    "            else:\n",
    "                sequence2sequence = scores[r-1][c-1] + mismatch\n",
    "                \n",
    "            # Compare scores to find the largest and the correct traceback(s)\n",
    "            if gap2sequence2 >= gap2sequence1 and gap2sequence2 >= sequence2sequence:\n",
    "                # left\n",
    "                scores[r][c] = gap2sequence2\n",
    "                traceback[r][c] = (r, c-1)\n",
    "            elif gap2sequence1 >= gap2sequence2 and gap2sequence1 >= sequence2sequence:\n",
    "                # up\n",
    "                scores[r][c] = gap2sequence1\n",
    "                traceback[r][c] = (r-1, c)\n",
    "            elif sequence2sequence >= gap2sequence1 and sequence2sequence >= gap2sequence2:\n",
    "                # diagonal\n",
    "                scores[r][c] = sequence2sequence\n",
    "                traceback[r][c] = (r-1, c-1)\n",
    "    \n",
    "    r = len(sequence1)\n",
    "    c = len(sequence2)\n",
    "    trace = traceback[r][c]\n",
    "    rev_seq1 = \"\"\n",
    "    rev_seq2 = \"\"\n",
    "    \n",
    "    while r != 0 or c != 0:\n",
    "        r_diff = trace[0] - r\n",
    "        c_diff = trace[1] - c\n",
    "        \n",
    "        if r_diff == -1 and c_diff == -1:\n",
    "            # diagonal trace\n",
    "            rev_seq1 += \"{}\".format(sequence1[r-1])\n",
    "            rev_seq2 += \"{}\".format(sequence2[c-1])\n",
    "            r += -1\n",
    "            c += -1\n",
    "        elif r_diff == -1:\n",
    "            # up trace\n",
    "            rev_seq1 += \"{}\".format(sequence1[r-1])\n",
    "            rev_seq2 += \"-\"\n",
    "            r += -1\n",
    "        elif c_diff == -1:\n",
    "            # left trace\n",
    "            rev_seq1 += \"-\"\n",
    "            rev_seq2 += \"{}\".format(sequence2[c-1])\n",
    "            c += -1\n",
    "            \n",
    "        trace = traceback[r][c]\n",
    "    \n",
    "    seq1_align = rev_seq1[::-1]\n",
    "    seq2_align = rev_seq2[::-1]\n",
    "    align_score = scores[-1][-1]\n",
    "        \n",
    "    return align_score, seq1_align, seq2_align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "d4050 is 1\n",
      "d4020 is 2\n",
      "d4030 is 3\n",
      "d4049 is 4\n",
      "d4029 is 5\n",
      "\n",
      "   1  2  3  4  5\n",
      "1 [0. 4. 5. 4. 7.]\n",
      "2 [4. 0. 1. 5. 3.]\n",
      "3 [5. 1. 0. 6. 4.]\n",
      "4 [4. 5. 6. 0. 8.]\n",
      "5 [7. 3. 4. 8. 0.]\n"
     ]
    }
   ],
   "source": [
    "# 2. Compute the Hamming distance between all pairs of sequences and report in a matrix\n",
    "\n",
    "# Constants\n",
    "match = 2\n",
    "mismatch = -1\n",
    "gap = -2\n",
    "\n",
    "# Read in the sequences\n",
    "header, d4050 = input_genome(\"small.fasta\", 1)\n",
    "header, d4020 = input_genome(\"small.fasta\", 2)\n",
    "header, d4030 = input_genome(\"small.fasta\", 3)\n",
    "header, d4049 = input_genome(\"small.fasta\", 4)\n",
    "header, d4029 = input_genome(\"small.fasta\", 5)\n",
    "\n",
    "sequences = [d4050, d4020, d4030, d4049, d4029]\n",
    "\n",
    "#for sequence in sequences:\n",
    "    #print(sequence)\n",
    "    #print()\n",
    "    \n",
    "distances = np.zeros((5, 5))\n",
    "\n",
    "for i in range(0, len(sequences)):\n",
    "    for j in range(i+1, len(sequences)):\n",
    "        align_score, seq1_align, seq2_align = globalign(sequences[i], sequences[j], match, mismatch, gap)\n",
    "        \n",
    "        hamming = 0\n",
    "        for char1, char2 in zip(seq1_align, seq2_align):\n",
    "            if char1 != '-' and char2 != '-':\n",
    "                if char1 != char2:\n",
    "                    hamming += 1\n",
    "                    \n",
    "        distances[i][j] = hamming\n",
    "        distances[j][i] = hamming\n",
    "        \n",
    "        \n",
    "print(\"d4050 is 1\")\n",
    "print(\"d4020 is 2\")\n",
    "print(\"d4030 is 3\")\n",
    "print(\"d4049 is 4\")\n",
    "print(\"d4029 is 5\")\n",
    "print()\n",
    "print(\"   1  2  3  4  5\")\n",
    "i = 1\n",
    "for row in distances:\n",
    "    print(i, row)\n",
    "    i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here are my answers, but matched up to the order of your matrix\n",
    "# I replaced my Ns with As, which is why the sequence 1 pairs have a discrepancy\n",
    "\n",
    "# 14532\n",
    "#104754\n",
    "#4x0865\n",
    "#5xx043\n",
    "#3xxx01\n",
    "#2xxxx0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3\\. See the attached problem3.pdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4\\. Installed. The files are downloaded as large.phylip.txt and large.fasta.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5\\. The dnadist file is large.phylip.dnadist. The raw tree files for the kitsch program are large.phylip.kitsch.outtree (description) and large.phylip.kitsch.outfile (tree visualization). The raw tree files for the neighbor program are large.phylip.neighbor.outtree (description) and large.phylip.neighbor.outfile (tree visualization). \n",
    "\n",
    "Both trees are broken into a rough structure of three subtrees made up of (70, 71, 72), (64, 65, 66, 67, 80), and (75, 76, 77, 78, 79). There are many details which make the trees very dissimilar, including scrambling of the overall structure and differences in the subtree orders and structures. 69, 74, and 81 are placed in completely different locations in each tree. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6\\. I ran Clustal Omega and MUSCLE. The alignment files are named large.ebi.clustal.alignmentfile and large.ebi.muscle.alignmentfile. There are also pdfs of the generated trees name large.ebi.clustal.tree and large.ebi.muscle.tree. \n",
    "\n",
    "The Phylip and EBI trees have vague similarities in two subtrees (70-74) and (75-79), but the orderings are different, and the third subtree is replaced by a cascade of nodes. On a quick glance I would say they are very different. \n",
    "\n",
    "The Clustal Omega and MUSCLE trees are almost identical to each other. There are only two differences, which appear in the connections between 67, 80, and 64. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
